experimental/LinearQuotients/cox_rings.jl

```diff
@@ -0,0 +1,606 @@
+          push!(hom_invars, dot(v[i, :], basisd))
+
+          # The A-degree of the new element is given by the corresponding eigenvalues
+          push!(degrees, A(ZZRingElem[ div(powers_of_zeta[e[k]], l_div_eldiv[k]) for k = 1:length(e) ]))
+
+          # We added a new generator, so the corresponding element of T is just
+          # the next variable.
+          push!(hom_basisd, gens(T)[length(hom_invars)])
+        end
+      end
+      V = vcat(V, v)
+    end
+
+    # Now translate basisd into hom_basisd to be able to translate invars in
+    # hom_invars.
+
+    # We only need to do this for elements of basisd which are elements of invars.
+    row_to_invar = Int[]
+    M = zero_matrix(K, 0, length(basisd))
+    for i = 1:length(basisd)
+      if !is_gen[i]
+        continue
+      end
+      # Find k with invars[k] == basid[i]
+      k = findfirst(isequal(1), C.exponent_vectors[basisd[i]])
+      push!(row_to_invar, k)
+
+      N = zero_matrix(K, 1, length(basisd))
+      N[1, i] = one(K)
+      M = vcat(M, N)
+    end
+
+    fl, sol = can_solve_with_solution(V, M, side = :left)
+    @assert fl
+
+    # The i-th row of sol gives the coefficient of invars[row_to_invar[i]] in
+    # the basis hom_basisd.
+    for i = 1:nrows(sol)
+      invars_in_hom_invars[invars[row_to_invar[i]]] = dot(sol[i, :], hom_basisd)
+    end
+  end
+
+  # Now grade T, build the maps, and move the relations from S to T
+
+  T = grade(T, degrees)[1]
+
+  # Wants the images to be homogeneous, so have to turn off the check.
+  StoT = hom(S, T, [ invars_in_hom_invars[f] for f in invars ], check = false)
+  relsT = elem_type(T)[ StoT(r) for r in gens(modulus(Q)) ]
+
+  # TODO: Funnily enough ideal(relsT) would raise an error about the generators
+  # not being homogeneous (which is correct, I still want to build this ideal)
+  Q, TtoQ = quo(T, ideal(T, relsT))
+  QtoR = hom(Q, Rgraded, [ Rgraded(f) for f in hom_invars ])
+
+  return Q, QtoR
+end
+
+################################################################################
+#
+#  Cox ring of a QQ-factorial terminalization (aka minimal model) of a
+#  linear quotient
+#
+################################################################################
+
+@doc Markdown.doc"""
+    cox_ring_of_qq_factorial_terminalization(L::LinearQuotient)
+
+Return the Cox ring of a QQ-factorial terminalization of the linear quotient `L`
+in a presentation as a graded affine algebra (`MPolyQuoRing`) and an injective
+map from this ring into a Laurent polynomial ring using the algorithm from
+[Yam18](@cite).
+"""
+function cox_ring_of_qq_factorial_terminalization(L::LinearQuotient; verbose::Bool = false)
+  # TODO: can we handle this without error (in a type-stable way)?
+  @assert !has_terminal_singularities(L) "Variety is already terminal"
+
+  G = group(L)
+  is_subgroup_of_sl(G) || throw(Hecke.NotImplemented())
+
+  # Compute the Cox ring of L itself
+  RVG, RVGtoR = cox_ring(L)
+  verbose && @info "Computed K[V]^[G, G]"
```

Comment:
I guess this would be a good use case for the `@vprint` macro (I am not asking for this to be changed in this PR! but you may want to look into that for future work on this or other code)

Reply:
I forgot/didn't know about this. Thank you!